### PR TITLE
feat: Add config-manager push secrets

### PR DESCRIPTION
**Pull request by dallinjsevy**

### CODE REVIEW

**Review by phalestrivir — CHANGES_REQUESTED** [review]
See my comments below for the requested changes

File: src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import { printMessage, verboseMessage } from '../../../utils/Console';
+import { FrodoCommand } from '../../FrodoCommand';
+
+const { CLOUD_DEPLOYMENT_TYPE_KEY, FORGEOPS_DEPLOYMENT_TYPE_KEY } =
+  frodo.utils.constants;
+
+const deploymentTypes = [
+  CLOUD_DEPLOYMENT_TYPE_KEY,
+  FORGEOPS_DEPLOYMENT_TYPE_KEY,
+];
```
Comment (phalestrivir) [inline src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts]: This is a cloud only command, so remove forgeops

File: src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        '-n, --name <name>',
+        'Secret name; import only the specified secret'
+      )
+    )
+    .addOption(
+      new Option(
+        '-e, --env <values>',
+        'Value to set for the secret. Will override .env files and environment variables.'
+      )
+    )
```
Comment (phalestrivir) [inline src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts]: Delete this since the flag will be global

File: src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
```
Comment (phalestrivir) [inline src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts:35]: Need an option for `-p, --prune`

File: src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      // unrecognized combination of options or no options
+      else {
+        printMessage(
+          'Unrecognized combination of options or no options...',
+          'error'
+        );
+        program.help();
+        process.exitCode = 1;
+      }
+    });
```
Comment (phalestrivir) [inline src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts]: Change this code to look something like this: https://github.com/trivir/frodo-cli/blob/2c9d195e2dd86f4484e22c411549587c05f8fe3e/src/cli/config-manager/config-manager-push/config-manager-push-email-provider.ts#L35-L43

Since we don't need to display the help menu here, just change the wording from email providers to secrets

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 5]
```diff
@@ -87,3 +94,123 @@ export async function configManagerExportSecrets(
...
+  let value: string;
+  if (name in envFile) {
+    value = envFile[name];
+  } else if (name in process.env) {
+    value = process.env[name];
+  } else {
+    throw new FrodoError(`No value found for ${name}`);
+  }
+  return isBase64 ? value : Buffer.from(value).toString('base64');
+}
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: Delete this, since library will handle it

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 6]
```diff
@@ -87,3 +94,123 @@ export async function configManagerExportSecrets(
...
+
+export async function configManagerImportSecrets(
+  secretName?: string,
+  value?: string
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: Remove this value parameter

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 7]
```diff
@@ -87,3 +94,123 @@ export async function configManagerExportSecrets(
...
+    value = envFile[name];
+  } else if (name in process.env) {
+    value = process.env[name];
+  } else {
+    throw new FrodoError(`No value found for ${name}`);
+  }
+  return isBase64 ? value : Buffer.from(value).toString('base64');
+}
+
+export async function configManagerImportSecrets(
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts:109]: Add a function comment

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 8]
```diff
@@ -87,3 +94,123 @@ export async function configManagerExportSecrets(
...
+    const secretsDir = getFilePath(`esvs/secrets/`);
+    if (!fs.existsSync(secretsDir)) {
+      stopProgressIndicator(spinnerId, `No secrets found`, 'fail');
+      return true;
+    }
+
+    const fileNames = fs
+      .readdirSync(secretsDir)
+      .filter((name) => name.toLowerCase().endsWith('.json'))
+      .filter((name) => !secretName || name === secretName);
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: This should be `.filter((name) => !secretName || name ===  `${secretName}.json`);`

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 9]
```diff
@@ -87,3 +94,123 @@ export async function configManagerExportSecrets(
...
+    'indeterminate',
+    0,
+    `Reading secrets...`
+  );
+  let indicatorId: string;
+  try {
+    const secretsDir = getFilePath(`esvs/secrets/`);
+    if (!fs.existsSync(secretsDir)) {
+      stopProgressIndicator(spinnerId, `No secrets found`, 'fail');
+      return true;
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: This should return false since it's an error

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 10]
```diff
@@ -87,3 +94,123 @@ export async function configManagerExportSecrets(
...
+
+    if (fileNames.length === 0) {
+      stopProgressIndicator(
+        spinnerId,
+        secretName
+          ? `No matching secret found for ${secretName}`
+          : 'No secrets found to import',
+        'fail'
+      );
+      return true;
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: This should return false since it's an error

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 11]
```diff
@@ -87,3 +94,123 @@ export async function configManagerExportSecrets(
...
+    }
+
+    stopProgressIndicator(
+      spinnerId,
+      `Successfully read ${fileNames.length} secrets.`,
+      'success'
+    );
+
+    const envFile = loadEnvFile();
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: Delete this line, it's not necessary

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 12]
```diff
@@ -87,3 +94,123 @@ export async function configManagerExportSecrets(
...
+    throw new FrodoError(`No value found for ${name}`);
+  }
+  return isBase64 ? value : Buffer.from(value).toString('base64');
+}
+
+export async function configManagerImportSecrets(
+  secretName?: string,
+  value?: string
+): Promise<boolean> {
+  const errors = [];
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: We don't need an errors array since you will want to update this to use importSecrets where it already handles each error separate

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 13]
```diff
@@ -87,3 +94,123 @@ export async function configManagerExportSecrets(
...
+  const spinnerId = createProgressIndicator(
+    'indeterminate',
+    0,
+    `Reading secrets...`
+  );
+  let indicatorId: string;
+  try {
+    const secretsDir = getFilePath(`esvs/secrets/`);
+    if (!fs.existsSync(secretsDir)) {
+      stopProgressIndicator(spinnerId, `No secrets found`, 'fail');
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts:122]: I'd update the description to say `No secrets directory found`

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 14]
```diff
@@ -87,3 +94,123 @@ export async function configManagerExportSecrets(
...
+        );
+      } catch (error) {
+        errors.push(error);
+      }
+    }
+    
+    if (errors.length > 0) {
+      throw new FrodoError(`Error importing secrets`, errors);
+    }
+    stopProgressIndicator(indicatorId, `${fileNames.length} secrets imported.`);
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: Similar to variables, we want to use `importSecrets`. That will look something like this:

```typescript
indicatorId = createProgressIndicator(
  'indeterminate',
  0,
  'Importing secrets'
);

const remoteSecrets = Object.fromEntries((await readSecrets()).map(s => [s._id, s]));
const secrets = {} as Record<string, ScriptSkeleton>;

for (const fileName of fileNames) {
  try {
    const secret = readJsonFile(`${secretsDir}/${fileName}`) as SecretSkeleton;
    if (prune) await pruneSecretVersions(secret._id, false, true);
    if (secret.valueBase64) {
      secret.activeValue = secret.valueBase64;
    } else {
       const versions = secret.versions.sort((a, b) => 
         Number(a.version) > Number(b.version) ? 1 : -1
       );
       for (let i = 0; i < versions.length -1; ++i) {
         if (i === 0 && !remoteSecrets[secret._id]) {
           await createSecret(
             secret._id,
             versions[i].valueBase64,
             secret.description,
             secret.encoding,
             secret.useInPlaceholders
            );
         } else {
           await createVersionOfSecret(secretId, versions[i].valueBase64);
         }
       }
       secret.activeValue = versions[versions.length - 1].valueBase64;
    }
    secrets[secret._id] = secret;
  } catch (e) {
     printError(e, `Error importing secret from "${fileName}"`);
  }
}

const imported = await importSecrets({ secret: secrets }, { includeActiveValues: true });

let unchanged = 0;
let updated = 0;

for (const s of imported) {
  if (remoteSecrets[s._id] && remoteSecrets[s._id].activeVersion === s.activeVersion) {
    printMessage(`Secret ${s._id} unchanged version ${s.activeVersion}`);
    unchanged++;
  } else {
    printMessage(`Secret ${s._id} created version ${s.activeVersion}`);
    updated++;
  }
}

stopProgressIndicator(
  indicatorId,
  `${imported.length} secrets imported.`
);

printMessage(
  updated > 0
    ? `Changes made to secrets: ${updated} updated, ${unchanged} unchanged`
    : `No changes, (${unchanged} secrets(s) already up to date)`
);
```

Note that you will need to add a `pruneSecretVersions` function to the library for the prune flag since we don't have a way to do this yet (we'll probably want to eventually create a separate Frodo command for this for normal Frodo, but adding it to the library should be fine for now). 

Also, since secrets are a bit more complicated that variables, I'm not certain the code above will be 100% correct, but I believe it is (but make sure to test it to be safe).

File: test/e2e/exports/fr-config-manager/cloud/esvs/secrets/esv-test-secret.json
Comment (phalestrivir) [inline test/e2e/exports/fr-config-manager/cloud/esvs/secrets/esv-test-secret.json:1]: I would update this one to instead have multiple versions, so we can test importing multiple versions. It should have at least two versions, but I'd probably do three to be safe.

File: test/e2e/mocks/default_2470140894/recording.har
Comment (phalestrivir) [inline test/e2e/mocks/default_2470140894/recording.har:1]: Delete this recording

File: test/e2e/config-manager-push-secrets.e2e.test.js [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * 5. Commit both your test and your new recordings to the repository.
+ *    Your tests are likely going to reside outside the frodo-lib project but
+ *    the recordings must be committed to the frodo-lib project.
+ */
+
+/*
+// ForgeOps
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://nightly.gcp.forgeops.com/am frodo config-manager push secrets -D test/e2e/exports/fr-config-manager/cloud 
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://nightly.gcp.forgeops.com/am frodo config-manager push secrets -n esv-fr-test-secret -e my-test-value -D test/e2e/exports/fr-config-manager/cloud
```
Comment (phalestrivir) [inline test/e2e/config-manager-push-secrets.e2e.test.js]: Since these are cloud tests, update the host to be the frodo-dev tenant

File: test/e2e/config-manager-push-secrets.e2e.test.js [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    test(`"frodo config-manager push secrets -D ${allDirectory} ": should import the secrets into cloud"`, async () => {
+        const CMD = `frodo config-manager push secrets -D ${allDirectory} `;
+        const { stdout, stderr } = await exec(CMD, {
+            env: {
+                ...cloudEnv.env,
+                ESV_FR_TEST_SECRET: "my-fr-test-secret-value",
+                ESV_TEST_SECRET: "my-test-secret-value"
+            }
+        });
+        expect(removeAnsiEscapeCodes(stdout)).toMatchSnapshot();
```
Comment (phalestrivir) [inline test/e2e/config-manager-push-secrets.e2e.test.js]: I would update this (and other tests) to use `testSuccess(CMD, cloudEnv);` You'll also need to update this particular test to have an env file flag to import all the secrets and versions.

File: test/e2e/mocks/config-manager_4167095917/push_2272264157/secrets_3084510534/0_D_m_314327836/am_1076162899/recording.har
Comment (phalestrivir) [inline test/e2e/mocks/config-manager_4167095917/push_2272264157/secrets_3084510534/0_D_m_314327836/am_1076162899/recording.har:1]: Delete this recording since it's not needed

File: test/e2e/mocks/config-manager_4167095917/push_2272264157/secrets_3084510534/0_D_m_314327836/environment_1072573434/recording.har
Comment (phalestrivir) [inline test/e2e/mocks/config-manager_4167095917/push_2272264157/secrets_3084510534/0_D_m_314327836/environment_1072573434/recording.har:1]: Delete this recording since it's not needed

File: test/e2e/mocks/config-manager_4167095917/push_2272264157/secrets_3084510534/0_D_m_314327836/oauth2_393036114/recording.har
Comment (phalestrivir) [inline test/e2e/mocks/config-manager_4167095917/push_2272264157/secrets_3084510534/0_D_m_314327836/oauth2_393036114/recording.har:1]: Delete this recording since it's not needed

**Review by phalestrivir — CHANGES_REQUESTED** [review]

File: src/configManagerOps/FrConfigSecretOps.ts (hunk 14 above)
Comment (phalestrivir) [reply src/configManagerOps/FrConfigSecretOps.ts]: @dallinjsevy I'm working on adding the `pruneSecretVersions` to Frodo lib, so no need to worry about that

Comment (phalestrivir) [reply src/configManagerOps/FrConfigSecretOps.ts]: Created PR for the `pruneSecretVersions`: https://github.com/rockcarver/frodo-lib/pull/624

File: esvs/secrets/esv-fr-test-secret.json
Comment (phalestrivir) [inline esvs/secrets/esv-fr-test-secret.json:1]: Delete this file

File: esvs/secrets/esv-test-secret.json
Comment (phalestrivir) [inline esvs/secrets/esv-test-secret.json:1]: Delete this file

File: src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+import { verboseMessage } from '../../../utils/Console';
+import { FrodoCommand } from '../../FrodoCommand';
+
+const { CLOUD_DEPLOYMENT_TYPE_KEY, FORGEOPS_DEPLOYMENT_TYPE_KEY } =
+  frodo.utils.constants;
+
+const deploymentTypes = [
+  CLOUD_DEPLOYMENT_TYPE_KEY,
+  FORGEOPS_DEPLOYMENT_TYPE_KEY,
+];
```
Comment (phalestrivir) [inline src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts]: This is a cloud only command, so remove forgeops

File: src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+  program
+    .description('Import secrets.')
+    .addOption(
+      new Option(
+        '-n, --name <name>',
+        'Secret name; import only the specified secret'
+      )
+    )
+
+    .addOption(new Option('-p, --prune', 'Prune old configureation'))
```
Comment (phalestrivir) [inline src/cli/config-manager/config-manager-push/config-manager-push-secrets.ts]: I'd change the description to indicate it prunes old active versions: `Prunes old versions that are still enabled before importing the new version(s)`

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 19]
```diff
@@ -87,3 +99,138 @@ export async function configManagerExportSecrets(
   }
   return false;
 }
+
+/**
+ * Import secrets to to cloud from fr-config-manager format
+ * @param {string} secretName optional name of secret to import
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts:105]: Missing prune description in the function comment

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 20]
```diff
@@ -87,3 +99,138 @@ export async function configManagerExportSecrets(
   }
   return false;
 }
+
+/**
+ * Import secrets to to cloud from fr-config-manager format
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: Remove duplicate "to" here

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 21]
```diff
@@ -87,3 +99,138 @@ export async function configManagerExportSecrets(
...
+          `${secretsDir}/${fileName}`
+        ) as SecretSkeleton & {
+          valueBase64?: string;
+          versions?: VersionOfSecretSkeleton[];
+        };
+        if (prune) await pruneVersionsOfSecret(secret._id, false, true);
+        if (secret.valueBase64 !== undefined) {
+          secret.activeValue = secret.valueBase64;
+        } else {
+          const versions = [...secret.versions].sort((a, b) =>
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: I don't think you need to do `[...secret.versions]` here, it's an extra step for copying that doesn't need to happen unless you don't want the original array sorted, but we don't care about the original because it shouldn't be used during imports. Instead just do `secret.versions.sort(...);`

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 22]
```diff
@@ -87,3 +99,138 @@ export async function configManagerExportSecrets(
...
+            Number(a.version) - Number(b.version) ? 1 : -1
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: This is wrong, you don't want to use `-` here, you want to use `>`. If you use `-`, the only time this is falsy is when the versions are equal, meaning you'd actually want to return `0` in that case. This is a comparator function, so you return 1 if the left value is greater than the right value, -1 if the left value is less than the right value, and 0 if they are equal (although if they are equal then the order doesn't really matter, so we don't handle the 0 case, but we could if we want to, but since there are no versions that will be the same we don't need to handle it).

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 23]
```diff
@@ -87,3 +99,138 @@ export async function configManagerExportSecrets(
...
+        }
+        printError(e, `Error importing secret from "${fileName}"`);
+      }
+    }
+    const includeActiveValues = true;
+
+    const imported = await importSecrets(
+      { secret: secrets },
+      includeActiveValues
+    );
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: We don't need a variable for includeActiveValues, I would just simplify this to:

```typescript
const imported = await importSecrets(
  { secret: secrets },
  includeActiveValues: true
);
```

File: test/client_cli/en/__snapshots__/config-manager-push-secrets.test.js.snap [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+                      or '/parent/child' otherwise. (default: "alpha" for
+                      Identity Cloud tenants, "/" otherwise.)
+  username            Username to login with. Must be an admin user with
+                      appropriate rights to manage authentication
+                      journeys/trees.
+  password            Password.
+
+Options:
+  -e, --env <values>  Value to set for the secret. Will override .env files and
+                      environment variables.
```
Comment (phalestrivir) [inline test/client_cli/en/__snapshots__/config-manager-push-secrets.test.js.snap]: This snapshot needs to be updated

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 25]
```diff
@@ -87,3 +99,138 @@ export async function configManagerExportSecrets(
...
+      spinnerId,
+      `Successfully read ${fileNames.length} secrets.`,
+      'success'
+    );
+
+    indicatorId = createProgressIndicator(
+      'indeterminate',
+      0,
+      'Importing secrets'
+    );
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts:152]: This is a whoops on my part with the code I originally provided, this should be a 'determinate' progress indicator:

```typescript
indicatorId = createProgressIndicator(
  'determinate',
  fileNames.length,
  'Importing secrets'
);
``

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 26]
```diff
@@ -87,3 +99,138 @@ export async function configManagerExportSecrets(
...
+        secrets[secret._id] = secret;
+      } catch (e) {
+        if (
+          e instanceof Error &&
+          e.message.includes('No value found for placeholder')
+        ) {
+          continue;
+        }
+        printError(e, `Error importing secret from "${fileName}"`);
+      }
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts:192]: At the end of this for-loop, we need to update the progress indicator indicating we are done importing a particular secret.

File: test/e2e/__snapshots__/config-manager-push-secrets.e2e.test.js.snap [hunk 27]
```diff
@@ -0,0 +1,43 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`frodo config-manager push secrets "frodo config-manager push secrets -D test/e2e/exports/fr-config-manager/cloud ": should import the secrets into cloud" 1`] = `""`;
+
+exports[`frodo config-manager push secrets "frodo config-manager push secrets -D test/e2e/exports/fr-config-manager/cloud ": should import the secrets into cloud" 2`] = `
+"Experimental feature in use: 'frodo config-manager push secrets'. This feature may change without notice.
+✔ Successfully read 4 secrets.
+• 0 secrets imported.
```
Comment (phalestrivir) [inline test/e2e/__snapshots__/config-manager-push-secrets.e2e.test.js.snap]: I would expect it to say that 4 secrets were imported since it read in 4 secrets. My guess is that you didn't set the ENVs (since I don't see any env flags), so they got skipped (that's the other problem I left a comment about, we should see the error messages for why they got skipped, but currently those are not being shown)

Update this test to set the envs before importing so that they actually get imported.

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 28]
```diff
@@ -87,3 +99,138 @@ export async function configManagerExportSecrets(
...
+          secret.activeValue = versions[versions.length - 1].valueBase64;
+        }
+        secrets[secret._id] = secret;
+      } catch (e) {
+        if (
+          e instanceof Error &&
+          e.message.includes('No value found for placeholder')
+        ) {
+          continue;
+        }
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts]: Delete this, we need to see the errors for the placeholders not found.

File: test/e2e/config-manager-push-secrets.e2e.test.js [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+ *    Your tests are likely going to reside outside the frodo-lib project but
+ *    the recordings must be committed to the frodo-lib project.
+ */
+
+/*
+// cloud
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets -D test/e2e/exports/fr-config-manager/cloud 
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets -n esv-fr-test-secret -E ESV_FR_TEST_SECRET="test1" -D test/e2e/exports/fr-config-manager/cloud 
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets -n esv-fr-test-secret --env ESV_FR_TEST_SECRET="new fr version" -n esv-test-secret --env ESV_TEST_SECRET="new test version"  -D test/e2e/exports/fr-config-manager/cloud 
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets -n esv-test-secret -E ESV_TEST_SECRET="tes6" -n esv-test-secret-2 -E ESV_TEST_SECRET="test7" -D test/e2e/exports/fr-config-manager/cloud
```
Comment (phalestrivir) [inline test/e2e/config-manager-push-secrets.e2e.test.js]: These two tests you added are not necessary (we don't need to test duplicate `n` flags since only one of them actually get used). Instead, what I would do is create two tests to replace these ones:

The first test should test with `-n` again, but use `esv-test-secret` instead of `esv-fr-test-secret`, and update `esv-test-secret` to have multiple versions. You can delete the ESVs you added, `esv-test-secret-2` and `esv-test-secret-3`, since we don't need them for testing.

The second test you should test with `-D` and `-p` to test that the pruning works.

File: test/e2e/exports/fr-config-manager/cloud/esvs/secrets/esv-test-secret-3.json
Comment (phalestrivir) [inline test/e2e/exports/fr-config-manager/cloud/esvs/secrets/esv-test-secret-3.json:1]: Delete this as per https://github.com/trivir/frodo-cli/pull/70/changes#r3823366213

File: test/e2e/exports/fr-config-manager/cloud/esvs/secrets/esv-test-secret-2.json
Comment (phalestrivir) [inline test/e2e/exports/fr-config-manager/cloud/esvs/secrets/esv-test-secret-2.json:1]: Delete this as per https://github.com/trivir/frodo-cli/pull/70/changes#r3823366213

File: test/e2e/exports/fr-config-manager/cloud/esvs/secrets/esv-test-secret.json
Comment (phalestrivir) [inline test/e2e/exports/fr-config-manager/cloud/esvs/secrets/esv-test-secret.json:1]: Update this ESV to have multiple versions. That will look something like this:

```typescript
{
  "_id": "esv-test-secret",
  "description": "This is a frodo test",
  "encoding": "generic",
  "useInPlaceholders": true,
  "versions": [
    {
      "valueBase64": "${ESV_TEST_SECRET_1}",
      "status": "DESTROYED",
      "version": "1"
    },
    {
      "valueBase64": "${ESV_TEST_SECRET_2}",
      "status": "DISABLED",
      "version": "2"
    },
    {
      "valueBase64": "${ESV_TEST_SECRET_3}",
      "status": "ENABLED",
      "version": "3"
    },
    {
      "valueBase64": "${ESV_TEST_SECRET_4}",
      "status": "DESTROYED",
      "version": "4"
    },
    {
      "valueBase64": "${ESV_TEST_SECRET_5}",
      "status": "DISABLED",
      "version": "5"
    },
    {
      "valueBase64": "${ESV_TEST_SECRET_6}",
      "status": "ENABLED",
      "version": "6"
    }
  ]
}
```

**Review by phalestrivir — CHANGES_REQUESTED** [review]

File: src/configManagerOps/FrConfigSecretOps.ts [hunk 30]
```diff
@@ -87,3 +99,130 @@ export async function configManagerExportSecrets(
...
+              await createVersionOfSecret(secret._id, versions[i].valueBase64);
+            }
+          }
+          secret.activeValue = versions[versions.length - 1].valueBase64;
+        }
+        secrets[secret._id] = secret;
+      } catch (e) {
+        printError(e, `Error importing secret from "${fileName}"`);
+      }
+      updateProgressIndicator(indicatorId, `Exported secret ${secrets._id}`);
```
Comment (phalestrivir) [inline src/configManagerOps/FrConfigSecretOps.ts:193]: A few things here. We should have two updates for progress indicator, one for sucess and one for failure (or one like you have is fine, but in the event of an error we probably should have an error message instead of a success message). Second, the message is wrong on this one, it should say imported, and also `secrets._id` won't work. It should look something like this:

```typescript
  secrets[secret._id] = secret;
  updateProgressIndicator(indicatorId, `Imported secret ${secret._id}`);
} catch (e) {
  const errorMessage = `Error importing secret from "${fileName}"`;
  printError(e, errorMessage);
  updateProgressIndicator(indicatorId, errorMessage);
}

```

File: test/e2e/config-manager-push-secrets.e2e.test.js [hunk 31]
```diff
@@ -0,0 +1,87 @@
...
+ *    Your tests are likely going to reside outside the frodo-lib project but
+ *    the recordings must be committed to the frodo-lib project.
+ */
+
+/*
+// cloud
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets --env-file test/e2e/env/configManager1.env  -D test/e2e/exports/fr-config-manager/cloud 
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets -n esv-fr-test-secret -E ESV_FR_TEST_SECRET="test1" -D test/e2e/exports/fr-config-manager/cloud 
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets -n esv-fr-test-secret --env ESV_FR_TEST_SECRET="new fr version" -n esv-test-secret --env ESV_TEST_SECRET="new test version"  -D test/e2e/exports/fr-config-manager/cloud 
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets -p --name esv-test-secret --env-file test/e2e/env/configManager1.env   -D test/e2e/exports/fr-config-manager/cloud
```
Comment (phalestrivir) [inline test/e2e/config-manager-push-secrets.e2e.test.js:54]: Delete this line, it's no longer a test in this file

File: test/e2e/config-manager-push-secrets.e2e.test.js [hunk 32]
```diff
@@ -0,0 +1,87 @@
...
+ * 5. Commit both your test and your new recordings to the repository.
+ *    Your tests are likely going to reside outside the frodo-lib project but
+ *    the recordings must be committed to the frodo-lib project.
+ */
+
+/*
+// cloud
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets --env-file test/e2e/env/configManager1.env  -D test/e2e/exports/fr-config-manager/cloud 
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets -n esv-fr-test-secret -E ESV_FR_TEST_SECRET="test1" -D test/e2e/exports/fr-config-manager/cloud 
+FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo config-manager push secrets -n esv-fr-test-secret --env ESV_FR_TEST_SECRET="new fr version" -n esv-test-secret --env ESV_TEST_SECRET="new test version"  -D test/e2e/exports/fr-config-manager/cloud
```
Comment (phalestrivir) [inline test/e2e/config-manager-push-secrets.e2e.test.js:53]: This test should be testing importing multiple versions from esv-test-secret. You need to remove the duplicate `-n` flag, and then also specify all version ENV values. You'll want to use the `--env-file` flag for that like you did in the first test.

**Review by phalestrivir — CHANGES_REQUESTED** [review]
Need a few more changes

File: test/e2e/mocks/config-manager_4167095917/push_2272264157/secrets_3084510534/0_n_E_n_E_D_602310692/am_1076162899/recording.har
Comment (phalestrivir) [inline test/e2e/mocks/config-manager_4167095917/push_2272264157/secrets_3084510534/0_n_E_n_E_D_602310692/am_1076162899/recording.har]: Delete this recording

File: test/e2e/mocks/config-manager_4167095917/push_2272264157/secrets_3084510534/0_n_E_n_E_D_602310692/environment_1072573434/recording.har
Comment (phalestrivir) [inline test/e2e/mocks/config-manager_4167095917/push_2272264157/secrets_3084510534/0_n_E_n_E_D_602310692/environment_1072573434/recording.har]: Delete this recording

**Review by phalestrivir — APPROVED** [review]
Squash your commits and create the PR against Rockcarver

**Comment by phalestrivir** [comment]
Make sure to pull my commits down that I added before squashing
